pkg/controller/vsphere/reconciler.go

```diff
@@ -1088,6 +1099,113 @@ func getDiskSpec(s *machineScope, devices object.VirtualDeviceList) (types.BaseV
+	primaryDisk := disks[0].(*types.VirtualDisk)
+
+	// Let's create the data disks now
+	additionalDisks := []types.BaseVirtualDeviceConfigSpec{}
+	for i := range s.providerSpec.DataDisks {
+		diskSpec := s.providerSpec.DataDisks[i]
+		klog.InfoS("Adding disk", "spec", diskSpec)
+
+		//  Get controller.  Only supporting using same controller as primary disk at this time
+		controller, ok := devices.FindByKey(primaryDisk.ControllerKey).(types.BaseVirtualController)
+		if !ok {
+			klog.Infof("Unable to get scsi controller")
+		}
+
+		dev := &types.VirtualDisk{
+			VirtualDevice: types.VirtualDevice{
+				Key: devices.NewKey() - int32(i),
+				Backing: &types.VirtualDiskFlatVer2BackingInfo{
+					DiskMode:        string(types.VirtualDiskModePersistent),
+					ThinProvisioned: types.NewBool(true),
+					VirtualDeviceFileBackingInfo: types.VirtualDeviceFileBackingInfo{
+						FileName:  "",
+						Datastore: types.NewReference(datastore.Reference()),
+					},
+				},
+				ControllerKey: controller.GetVirtualController().Key,
+			},
+			CapacityInKB: int64(diskSpec.SizeGiB * 1024 * 1024),
+		}
+
+		assignUnitNumber(dev, devices, additionalDisks, controller, unit)
+		unit = *dev.UnitNumber
+
+		diskConfigSpec := types.VirtualDeviceConfigSpec{
+			Device:        dev,
+			Operation:     types.VirtualDeviceConfigSpecOperationAdd,
+			FileOperation: types.VirtualDeviceConfigSpecFileOperationCreate,
+		}
+
+		klog.V(1).InfoS("Generated device", "disk", dev)
+
+		additionalDisks = append(additionalDisks, &diskConfigSpec)
+	}
+	diskSpecs = append(diskSpecs, additionalDisks...)
+
+	return diskSpecs, nil
+}
+
+// assignController assigns a device to a controller.
+func assignUnitNumber(device types.BaseVirtualDevice, existingDevices object.VirtualDeviceList, newDevices []types.BaseVirtualDeviceConfigSpec, controller types.BaseVirtualController, offset int32) {
+	vd := device.GetVirtualDevice()
+	vd.ControllerKey = controller.GetVirtualController().Key
+	vd.UnitNumber = &offset
+
+	units := make([]bool, 30)
+	for i := 0; i < int(offset); i++ {
+		units[i] = true
+	}
+
+	switch sc := controller.(type) {
+	case types.BaseVirtualSCSIController:
+		//  The SCSI controller sits on its own bus
+		units[sc.GetVirtualSCSIController().ScsiCtlrUnitNumber] = true
+	}
+
+	key := controller.GetVirtualController().Key
+
+	// Check all existing devices
+	for _, device := range existingDevices {
+		d := device.GetVirtualDevice()
+
+		if d.ControllerKey == key && d.UnitNumber != nil {
+			units[int(*d.UnitNumber)] = true
+		}
+	}
+
+	// Check new devices
+	for _, device := range newDevices {
+		d := device.GetVirtualDeviceConfigSpec().Device.GetVirtualDevice()
+
+		if d.ControllerKey == key && d.UnitNumber != nil {
+			units[int(*d.UnitNumber)] = true
+		}
+	}
+
+	for unit, used := range units {
+		if !used {
+			unit32 := int32(unit)
+			vd.UnitNumber = &unit32
+			klog.Infof("Assigned unit number %d = %d", unit, *vd.UnitNumber)
```

Comment:
Wondering if we are getting a little too much on the logs here? This seems rather low level for logging default